packages/yarnpkg-pnpify/sources/NodeModulesFS.ts

```diff
@@ -34,15 +56,50 @@ type PortableNodeModulesFSOptions = {
   baseFs?: FakeFS<PortablePath>
 };
 
+type WatchedDirMap = Map<PortablePath, WatchedDir>;
+
+interface WatchedDir {
+  lastWatcherId: number;
+  watchers: { [watcherId: number]: Watcher & EventEmitter };
+  entries: Filename[];
+}
+
 class PortableNodeModulesFs extends FakeFS<PortablePath> {
   private readonly baseFs: FakeFS<PortablePath>;
-  private readonly pathResolver: NodePathResolver;
+  private readonly watchedDirs: WatchedDirMap = new Map();
+  private pathResolver: NodePathResolver;
 
   constructor(pnp: PnpApi, {baseFs = new NodeFS()}: PortableNodeModulesFSOptions = {}) {
     super(ppath);
 
     this.baseFs = baseFs;
     this.pathResolver = new NodePathResolver(pnp);
+    const pnpFilePath = npath.join(pnp.getPackageInformation(pnp.topLevel)!.packageLocation, '.pnp.js');
+    this.watchPnpFile(pnpFilePath);
+  }
+
+  private watchPnpFile(pnpFilePath: NativePath) {
+    this.baseFs.watch(NodeFS.toPortablePath(pnpFilePath), () => {
+      delete require.cache[pnpFilePath];
+      const pnp = require(pnpFilePath);
+      // Sometimes pnp object is empty, which is weird, but eventually it comes correctly filled
+      if (Object.keys(pnp).length > 0) {
+        this.pathResolver = new NodePathResolver(pnp);
+
+        for (const [watchPath, watchedDir] of this.watchedDirs) {
+          const newEntries = this.resolvePath(watchPath).dirList || [];
+          // Difference between new and old directory contents
+          const entryDiff = newEntries.filter(entry => !watchedDir.entries.includes(entry))
+            .concat(watchedDir.entries.filter(entry => !newEntries.includes(entry)));
```

Comment:
I'm a bit concerned by the runtime complexity 🤔 Would making `entries` a `Set` help?